dev/SampleFlatC/SampleFlatC.cpp

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "SampleFlatC.h"
+#include <string>
+
+struct MyRealType
+{
+    std::wstring theString;
+};
+
+STDAPI SampleFlatApiCreate(
+    _In_ LPCWSTR theString,
+    _Outptr_ HSAMPLEFLATAPI* sampleFlat) try
+{
+    RETURN_HR_IF(E_INVALIDARG, !theString);
+    auto real = std::make_unique<MyRealType>();
+    real->theString = theString;
+    *sampleFlat = reinterpret_cast<HSAMPLEFLATAPI>(real.release());
+    return S_OK;
+}
+CATCH_RETURN();
```

Comment:
Can we encourage a noexcept approach from the get-go? the try/catch approach often leads to leaking exceptions into HRs that don't mean anything to callers, as well as hard-to-diagnose crashes.

Reply:
Sadly the ABI needs some way to return failures to the caller. This is just one small example - maybe `theString` was not a palindrome, or only had an even number of the letter "e".  Failfasting on those seems problematic. This is also the "flat C" version, around which would eventually be wrapped metadata and then a language specific projection.  So C# would see something like:

```c#
ProjectReunion.SampleFlatApi.Instance a = ProjectReunion.SampleFlat.Create("kittens");
var s = a.String;
```

... which would throw `SAMPLEFLAT_E_ODD_NUMBER_OF_E` on the first line.

Reply:
Argument validation can return errors but everything else should pretty much failfast. And argument checking should be happening pretty near the API entrypoint. We don't have to block this change on it but we need a clear story here before we start accumulating more code.